adding a note to the documentation of polar

NivekT opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #63259

Fix #52919

Differential Revision: [D30342536](https://our.internmc.facebook.com/intern/diff/D30342536)

NivekT committed:
adding a note to documentation of polar

[ghstack-poisoned]

NivekT commented:
Note that we still need to decide if we will add a check to prevent negative inputs arguments. If not, then the current change should be sufficient.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -2360,6 +2360,10 @@ def merge_dicts(*dicts):
 
+.. note::
+    This implementation follows C++'s `std::polar <https://en.cppreference.com/w/cpp/numeric/complex/polar>`_
+    which is different from Python's `cmath.polar` and SciPy's `linalg.polar`.
```
How are they different?

anjali411 replied:
`cmath.polar` and  SciPy's `linalg.polar` are both the inverse of `torch.polar` (they in a complex number and returns `(r, theta)`)

NivekT replied:
I will add a sentence about their differences in the doc.

anjali411 replied:
yeah btw I don't think we usually reference the std functions for the Python API so I am not sure if we should reference `std::polar` especially since the standard library functions don't work for complex numbers on CUDA (we call into thrust library!)

mruberry replied:
When we're wrapping a standard library call it's not weird to reference it, like fmod:

https://pytorch.org/docs/master/generated/torch.fmod.html

The standard library is well documented.

As for CUDA, that's a good technical point but as long as the polar function we're calling there is supposed to be std::polar I think it's OK. As pointed out in your comment below, however, we don't actually call std::polar!

anjali411 replied:
> When we're wrapping a standard library call it's not weird to reference it, like fmod:

ahh ok!

mruberry commented on torch/_torch_docs.py:
```diff
@@ -2360,6 +2360,10 @@ def merge_dicts(*dicts):
 :attr:`angle`.
 
+.. note::
+    This implementation follows C++'s `std::polar <https://en.cppreference.com/w/cpp/numeric/complex/polar>`_
```
"follows" is vague - how is torch.polar like C++'s std::polar?

mruberry replied:
(does torch.polar always call std::polar?)

anjali411 replied:
we do follow the same logic as `std::polar`, however we don't call into `std::polar`. 
https://github.com/pytorch/pytorch/blob/a9b0a921d592b328e7e80a436ef065dadda5f01b/aten/src/ATen/native/cpu/ComplexKernel.cpp#L18-L24 I don't see a reason why we shouldn't though!

NivekT replied:
I am updating the description to say that this function follows the exact logic of `std::polar` and that the input is unvalidated (i.e. it will still perform that computation as described even if the `abs` input tensor contains negative values`.

mruberry commented:
+ @anjali411, our complex expert

anjali411 reviewed (commented):
I am not sure about the utility of this note, especially because the function description is straightforward and clear. I think these notes are useful when we differ from libraries like numpy or SciPy on subtle details, but that's not the case here.

On a different note, we might wanna consider being in line with SciPy and changing the name of this operation.

NivekT committed:
Update on "adding a note to the documentation of polar"


Fix #52919

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30342536).

NivekT commented:
I can see this note being useful for users of `cmath` and `SciPy` who expect this function to work the same way as it does in the other two libraries, but doesn't know about the implementation of `std::polar` exists. 

I do agree that it may be better to follow `SciPy`'s implementation, but that may have backward compatibility issue?

anjali411 commented:
> Note that we still need to decide if we will add a check to prevent negative inputs arguments. If not, we should make it explicit that negative values are allowed.

`std::polar` doesn't error out for negative abs value so I guess we can do the same too.

anjali411 commented on torch/_torch_docs.py:
```diff
@@ -2360,12 +2360,20 @@ def merge_dicts(*dicts):
 corresponding to the polar coordinates with absolute value :attr:`abs` and angle
 :attr:`angle`.
```
```suggestion
:attr:`angle`. The behavior is undefined if r is negative or NaN, or if theta is infinite.
```

NivekT replied:
Added

anjali411 commented on torch/_torch_docs.py:
```diff
@@ -2360,12 +2360,20 @@ def merge_dicts(*dicts):
-    abs (Tensor): The absolute value the complex tensor. Must be float or
-        double.
+    abs (Tensor): The absolute value the complex tensor. Must be float or double.
+        Note: this input is not validated, and the function will still perform the computation as described
```
let's move the extra detail to the function description above.

NivekT replied:
Done

NivekT committed:
Update on "adding a note to the documentation of polar"


Fix #52919

Differential Revision: [D30342536](https://our.internmc.facebook.com/intern/diff/D30342536)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30342536).

mruberry commented on torch/_torch_docs.py:
```diff
@@ -2362,10 +2362,22 @@ def merge_dicts(*dicts):
 .. math::
     \text{out} = \text{abs} \cdot \cos(\text{angle}) + \text{abs} \cdot \sin(\text{angle}) \cdot j
+
+The behavior of this function is undefined if the absolute value is NaN, or if the angle is infinite.
```
I would move these paragraphs into the note

NivekT replied:
Done

mruberry commented on torch/_torch_docs.py:
```diff
@@ -2362,10 +2362,22 @@ def merge_dicts(*dicts):
+because that is the behavior defined by `std::polar`.
+
+.. note::
+    This implementation has the same exact logic as C++'s implementation of
```
Suggestion: 

"torch.polar is similar to std::polar (link) and does not compute the polar decomposition of a complex tensor like Python's cmath.polar and SciPy's linalg.polar do. Like std::polar, this operation does not validate its inputs."

NivekT replied:
I followed this suggestion.

NivekT committed:
Update on "adding a note to the documentation of polar"


Fix #52919

Differential Revision: [D30342536](https://our.internmc.facebook.com/intern/diff/D30342536)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30342536).

NivekT commented:
I think it should be good to go now. Please have a look and approve if you agree.

anjali411 commented on torch/_torch_docs.py:
```diff
@@ -2362,10 +2362,18 @@ def merge_dicts(*dicts):
+    `std::polar <https://en.cppreference.com/w/cpp/numeric/complex/polar>`_
+    and does not compute the polar decomposition
+    of a complex tensor like Python's `cmath.polar` and SciPy's `linalg.polar` do.
+    Like `std::polar`, this operation does not validate its inputs.
+    The behavior of this function is undefined if the absolute value is NaN, or if the angle is infinite.
```
```suggestion
    Like `std::polar`, this operation does not validate its inputs, i.e., the behavior of this function is undefined if `abs` is negative or NaN, or if `angle` is infinite.
```

NivekT replied:
The function still produces outputs when `abs` is negative, correct? If so, it is not undefined?

NivekT replied:
Spoke with @anjali411 offline. We will add this edit.

anjali411 replied:
actually can you just modify this sentence to just be: `The behavior of this function is undefined if `abs` is negative or NaN, or if `angle` is infinite.`, i.e., get rid of `Like `std::polar`, this operation does not validate its inputs,` ...

NivekT replied:
Yes

anjali411 reviewed (approved):
Thanks @NivekT for this doc update! I left one last comment, but the PR is good to go after that's addressed.

NivekT committed:
Update on "adding a note to the documentation of polar"


Fix #52919

Differential Revision: [D30342536](https://our.internmc.facebook.com/intern/diff/D30342536)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30342536).
